### PR TITLE
feature/15361-stock-cumulative-sum

### PR DESCRIPTION
**Pull request by raf18seb**
Added new feature [Cumulative Sum](https://www.highcharts.com/docs/stock/cumulative-sum). See #15361

### CODE REVIEW

File: ts/Core/Chart/StockChart.ts
Comment (karolkolodziej) [inline ts/Core/Chart/StockChart.ts:1038]: Isn't it always `true`?

File: ts/Core/Chart/StockChart.ts
Comment (karolkolodziej) [inline ts/Core/Chart/StockChart.ts:1234]: Isn't it always `true`?

File: ts/Core/Chart/StockChart.ts
Comment (karolkolodziej) [inline ts/Core/Chart/StockChart.ts:1309]: What do you mean by `after render time`?

**Review by karolkolodziej — CHANGES_REQUESTED** [review]

File: ts/Core/Chart/StockChart.ts
Comment (raf18seb) [reply ts/Core/Chart/StockChart.ts:1038]: `H.pick(false, true)` returns `false`

File: ts/Core/Chart/StockChart.ts
Comment (raf18seb) [reply ts/Core/Chart/StockChart.ts:1234]: `H.pick(false, true)` returns `false`

**Review by pawellysy — APPROVED** [review]
Good work!

File: ts/Core/Chart/StockChart.ts
Comment (pawelfus) [inline ts/Core/Chart/StockChart.ts:928]: > This method is called both from the axis
> * when finding dataMin and dataMax
 
 This is a little bit off. For compare it's correct, but not for `cumulativeSum`

**Review by pawelfus — CHANGES_REQUESTED** [review]

File: samples/stock/members/axis-setcumulative/demo.js
Comment (raf18seb) [reply samples/stock/members/axis-setcumulative/demo.js:8]: Q1 answer: it has never worked with compare too. I should fix it for both cumulative and compare then

Q2: sure 👍

File: ts/Core/Chart/StockChart.ts
Comment (raf18seb) [reply ts/Core/Chart/StockChart.ts:1309]: > What do you mean by after render time?
We should ask the author of the compare :) I think he meant that we don't have to define compare on the chart's init in the chart's options, but we can turn it on later - "after the chart has been initially rendered".

File: ts/Core/Chart/StockChart.ts
Comment (raf18seb) [reply ts/Core/Chart/StockChart.ts:1322]: Yes we can. I see 2 disadvantages of it:
- a little bit more difficult API for users that are using only one of compare/cumulative
- no backward compatibility for compare

What's your final decision? Leave it as it is or optimize the code?

**Review by pawelfus — APPROVED** [review]
Approved, but I'm thinking about refactoring the whole "compare"/"cumulative sum" modes. This extension is still overwriting some of the prototype methods (`Series.init`, `Series.processData` etc) and adding extra ones (`Series.initCompare`, `Series.setCompare` etc, however `Axis.setCompare()` should remain, as this is part of the official API). 

Instead, we could:
- use events instead of overwriting
- move to a separate file, make this an extension, e.g. `Series.dataModify = { init, set, others... }`

So we could use it later in tree shaking (extra ~400lines, not every Stock user needs these modes).

Refactoring of course would be in a separate PR/branch

What do you think @raf18seb @TorsteinHonsi @bre1470

**Comment by bre1470** [comment]
I like the second idea, a composition, it has the least impact on the axis and series itself. The official functions could be a link to the composition instance. The naming though needs improvement, `dataModify` is to unspecific. As this is a regular composition, it should go directly under the `ts/Series` folder.

File: ts/Core/Chart/StockChart.ts
Comment (bre1470) [reply ts/Core/Chart/StockChart.ts:1248]: It depends. You should consider a regular object with `as const` cast.
- Example: https://www.typescriptlang.org/play?ts=3.9.7#code/KYOwrgtgBAKgygUXNA3gKCpqID2AaDLCAQwE8AjYArKU4AZzQF800BjHEegFygClEyKAF4o6GrgBcUAAzUiZStICM8zHXrSATMyjF6UDl24BuNEA
- Docs: https://www.typescriptlang.org/docs/handbook/enums.html

File: ts/Core/Chart/StockChart.ts
Comment (karolkolodziej) [inline ts/Core/Chart/StockChart.ts:959]: I think there is no handler for the case when `modeState: null` 

`typeof modeState === 'object'`

File: ts/Core/Chart/StockChart.ts
Comment (karolkolodziej) [inline ts/Core/Chart/StockChart.ts:1035]: From what I see, casting to `any` isn't necessary.

File: ts/Core/Chart/StockChart.ts
Comment (karolkolodziej) [inline ts/Core/Chart/StockChart.ts:1308]: `defined(value) && defined(index)`

File: ts/Core/Chart/StockChart.ts
Comment (karolkolodziej) [reply ts/Core/Chart/StockChart.ts:1248]: Both `enum` and casting `as const` sound better than having this as a hardcoded string.
But casting `as const`  looks better after compiling that is true

**Review by karolkolodziej — CHANGES_REQUESTED** [review]

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Supporting a refactoring as suggested by Pawel. In addition:
- Add docs
- My name is in the author list in the samples. Author should be removed or updated to the correct name.

**Comment by pawelfus** [comment]
> * Add docs

Related PR: https://github.com/highcharts/highcharts/pull/16187

**Review by TorsteinHonsi — APPROVED** [review]
Thanks! Pending approval by @karolkolodziej

**Comment by raf18seb** [comment]
@TorsteinHonsi sorry for the number of merges, I had some troubles with a project. This PR still needs some polishing.

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Pending more polishing as stated above.

**Review by pawelfus — APPROVED** [review]
Thanks! 🎉

File: samples/stock/demo/cumulative-sum/demo.details
Comment (karolkolodziej) [inline samples/stock/demo/cumulative-sum/demo.details:6]: Is this a proper date? @pawelfus

**Review by karolkolodziej — APPROVED** [review]

File: samples/stock/demo/cumulative-sum/demo.details
Comment (pawelfus) [reply samples/stock/demo/cumulative-sum/demo.details:6]: Yes, it is 👍

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**
